pyrefly/lib/solver/subset.rs

```diff
@@ -932,12 +932,21 @@ impl<'a, Ans: LookupAnswer> Subset<'a, Ans> {
                     &want,
                 )
             }
-            (t1, Type::Quantified(q)) => match q.restriction() {
-                // This only works for constraints and not bounds, because a TypeVar must resolve to exactly one of its constraints.
-                Restriction::Constraints(constraints) => all(constraints.iter(), |constraint| {
-                    self.is_subset_eq(t1, constraint)
-                }),
-                _ => Err(SubsetError::Other),
+            (t1, Type::Quantified(q)) => {
+                // Allow 'Any' to match any TypeVar
+                if let Type::ClassType(t1_class) = t1 {
+                    if self.type_order.extends_any(t1_class.class_object()) {
+                        return Ok(());
+                    }
+                }
+                match q.restriction() {
+                    Restriction::Constraints(constraints) => all(constraints.iter(), |constraint| {
+                        self.is_subset_eq(t1, constraint)
+                    }),
+                    // Handle 'Unrestricted' and 'Bound' TypeVars
```

Comment:
This is wrong, the comment that was deleted explains why

The added logic would allow programs like this to type check

```
def [T]foo(x: T) -> T:
  x = 1  # this would be allowed by the Restriction::Unrestricted branch
  return x
  
y: str = foo("") # expect this to be str, but it's actually int
```

The reason it's safe to do for constrained type vars is because they must resolve to _exactly_ one of the constraints

so if you have something constrained by types A and B, it can be matched by a value that is both type A _and_ type B.

whereas when you have a bounded or unrestricted type var, it could be any type, so you can't be sure that a concrete type matches whatever it can resolve to.